### DISCUSSION TITLE
delay(cancelToken, ms) argument order

### DISCUSSION THREAD

**Discussion by noseratio** [esfx/esfx#82] [Ideas]
Hi Ron, first of all, many thanks for this library! Obviously, an enormous amount of time and efforts has gone into maintaining it. And for some of us who's switching back and forth between TypeScript and C#, the functionality it provides is simply indispensable.

I've just bumped into an error where I was using `await delay(ms, cancelToken)` instead of `await delay(cancelToken, ms)`, and it took me a good hour to figure out why the delay wasn't getting cancelled. Of course it was because this override was happily accepting `cancelToken` as the 2nd argument:

`export declare function delay<T>(msec: number, value: T | PromiseLike<T>): Promise<T>;`

I wonder if could do something about it, like maybe give it a runtime check inside `delay`? 

In .NET, the cancellation token is always passed as a last argument by convention (and `Task.Delay(ms, token)` follows that). That's how I got there in the first place :)  Is that's OK by you, I could submit a PR.

Thanks!

**Comment by rbuckton** [comment]
.NET's `Task.Delay` does not allow you to produce a result, but this version of `delay` does. There's no reason the result of a `delay` couldn't be a cancellation token, which is why I opted to make the cancellation token the first argument in the overload list so that you could, for example, write `delay(cancelToken1, ms, cancelToken2)`. 

That said, I could perhaps make it an error to pass a cancellation token for the `value` argument unless you call `delay` with `undefined` for the first argument, i.e.: `delay(undefined, ms, cancelToken)`. This would be a more explicit indication that you actually intended for the `Promise` to resolve with a cancellation token.

It would be feasible to indicate this in the type system as well, if a bit awkward, e.g.:

```ts
export declare function delay<T>(msec: number, value: T): T extends Cancelable ? never : Promise<Awaited<T>>;
```

**Reply by noseratio** [reply]
> That said, I could perhaps make it an error to pass a cancellation token for the value argument unless you call delay with undefined for the first argument

That'd an ideal outcome, thanks much Ron!

**Reply by rbuckton** [reply]
This change is available in `v1.1.0-alpha.202307260322` and will be in the next full release once I wrap up a few other small changes.

**Reply by noseratio** [reply]
Thank you very much! 👍
